### ISSUE TITLE
[BUG] Scala3 Schema.derived with recursive structures causes Deadlock with given keyword

### ISSUE THREAD

**Issue by hamnis** [softwaremill/tapir#2749]
Tapir version: 1.2.8

Scala version: ***

**Describe the bug**

When using Schema.derived on a recursive data structure, a deadlock is caused runtime.
It compiles fine, but fails runtime.

**How to reproduce?**

Minimized example:
https://gist.github.com/hamnis/eee66ef4189cca09b08505bb00d5406a

**Additional information**

Workaround, use `implicit def` for this

**Comment by adamw** [comment]
Yes, this is a known problem: https://tapir.softwaremill.com/en/latest/endpoint/schemas.html#derivation-for-recursive-types-in-scala3

Unfortunately I didn't write in the docs *why* this is the case. So probably worth investigating and properly documenting what is wrong.

**Comment by susliko** [comment]
Spent a few hours debugging a stalled app today.

### Source of the problem 

Code like this:
```scala
case class Animal(
  cub: Option[Animal]
) derives Schema
```
blocks a thread during initialization on `java.util.concurrent.CountDownLatch.await`.

That latch is a used by Scala lazy val initialization [mechanism](https://github.com/lampepfl/dotty/blob/6efcdbaa3df8a6b1e48dfcb208efe682bf5382e6/library/src/scala/runtime/LazyVals.scala#L55) to ensure some safety.

The blocking on the latch happens due to this line https://github.com/softwaremill/tapir/blob/5cd98a0e9773e1a3c55f81fe8f5e07fa9654fb10/core/src/main/scala-3/sttp/tapir/generic/auto/SchemaMagnoliaDerivation.scala#L41

Under the hood `p.typeclass` tries to reenter this same derivation process when `p` is `Option[Animal]`. There is a "cache" to avoid such loops, but it's not used for case class parameter derivation.

### What could be done

1. Having the `Mirror` instance we can get a full qualified type names of all case class parameters. Analyze these names and maybe hard-code cases for Option/List/etc. 
2. Write derivation in pure macros instead of Magnolia

**Comment by adamw** [comment]
Yeah, unfortunately that's a "known issue" (hence the documentation entry above). 

The best quick solution would be to solve this at the level of Magnolia - in a generic way, not specialised to Option/List.

I did try to derive schemas directly [once](https://github.com/softwaremill/tapir/pull/2118) but it proved more challenging that I though initially - at some point I got to rewriting Magnolia, which seemed to miss the point ;). But it might worthwhile revisiting the problem at some point, especially that Scala 3 derivation/macros matured somewhat.
